_drafts/unit-testing-basics.md

```diff
@@ -0,0 +1,1257 @@
+```
+
+To build and run this unit test, we give the unit test harness the test name,
+the list of files to compile into the test binary, and any extra compilation
+flags necessary.
+
+```
+COMPONENT_NAME=sum
+
+SRC_FILES = \
+  $(PROJECT_SRC_DIR)/my_sum.c \
+
+TEST_SRC_FILES = \
+  $(UNITTEST_SRC_DIR)/test_my_sum.c
+```
+
+Here, we have `SRC_FILES`, which would contain any sources files used by the
+test, and `TEST_SRC_FILES` which contains the test files that contain the tests
+themselves.
+
+## Unit Testing Best Practices
+
+The "Minimal Example" is a contrived example and very rarely will there be a
+test with _no_ other dependencies. Firmware is naturally coupled with other
+parts of hardware, and that makes it difficult at first to set up a unit test.
+
+For example, a flash storage module may call an `analytics_inc()` function to
+record the number of writes, a `watchdog_feed()` function during a large flash
+erase operation, and `timer_schedule()` to help defragment the flash later in
+the future. If we are testing only the flash K/V store, We _do not_ want to
+include the analytics, watchdog, and timer source files into our unit test.
+
+That brings us to a few best practices to follow, especially when writing unit
+tests for complex and entangled code.
+
+- Each `TEST()` within a unit test file should ideally test a single path or
+  feature of the module. A test called `TestEverything` is an anti-pattern.
+- Each test should be _quick_. A few milliseconds is ideal, and one second is
+  the worst case run time.
+- Each unit test should ideally include one _real_ implementation of a module.
+  The rest should be stubbed or fake versions of the modules not under test.
+- Those stubbed and fake versions of modules should be written early, reused,
+  and shared.
+
+Which brings us to explaining what are stubs, fakes, and mocks?
+
+## Stubs, Fakes, and Mocks
+
+When starting to write unit tests, it is common to write alternate
+implementations to modules that make sense for a particular unit test.
+
+Since unit tests will be run on the host machine, they won't have hardware, such
+as an LED. But if a module within a unit test calls `Enable_LED()`, we could
+instead have a virtual LED and a state boolean value saving whether the LED is
+on or off.
+
+These alternate implementations of modules have different types. Let's explain
+them.
+
+- **Fakes** will be a working implementation, but will usually substitute its
+  dependencies with something simpler and easier for a test environment.
+  Example: An in-memory key/value store vs A NOR Flash backed Key/Value store.
+- **Stubs** will be an implementation that returns canned values, generally
+  always returning valid or invalid values.
+- **Mocks** are an implementation that is controlled by the unit test. They can
+  be pre-programmed with return values, check values of arguments, and help
+  verify that functions are called.
```

Comment:
Fakes are often also controlled by the unit test? I.e. your fake time example below "allows incrementing the time manually". IMHO the distinction between mock vs fake is pretty small and mostly comes down to that mocks are usually directly involved in making assertions on when and how they get called (as you wrote). Because of that mocks are usually part of the unit test framework whereas fakes are usually not? Just my 2 cents.

Reply:
Fake time is definitely the weakest example in the list. I'll actually remove it. 

I don't think there is honestly much similarity between fakes and mocks. Fakes have logic, init, denit, state, etc. 

Mocks have none of that and are just glorified counters, validators and return value buckets.